src/api.ts

```diff
@@ -0,0 +1,410 @@
+import * as _ from "lodash";
+import * as querystring from "querystring";
+import * as requestModule from "request";
+import * as url from "url";
+import * as FormData from "form-data";
+
+import { Constants } from "./emulator/constants";
+import { FirebaseError } from "./error";
+import * as logger from "./logger";
+import * as responseToError from "./responseToError";
+import * as scopes from "./scopes";
+import * as utils from "./utils";
+
+export enum HttpMethod {
+  GET = "GET",
+  PUT = "PUT",
+  POST = "POST",
+  DELETE = "DELETE",
+  PATCH = "PATCH",
+}
+
+interface UploadFile extends FormData.AppendOptions {
+  stream: unknown;
+}
+
+interface FirebaseRequestLogOptions {
+  skipQueryParams?: boolean;
+  skipRequestBody?: boolean;
+  skipResponseBody?: boolean;
+}
+
+interface FirebaseRequestOptions {
+  auth?: boolean;
+  data?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  files?: { [k: string]: UploadFile };
+  form?: { [k: string]: any }; // eslint-disable-line @typescript-eslint/no-explicit-any
+  headers?: requestModule.Headers;
+  json?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  logOptions?: FirebaseRequestLogOptions;
+  origin: string;
+  qs?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  query?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  resolveOnHTTPError?: boolean;
+  retryCodes?: number[];
+  timeout?: number;
+}
+
+interface FirebaseResponse {
+  status: number;
+  response: requestModule.Response;
+  body: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```

Comment:
Is there anything at all we can say about the body?

Reply:
Shouldn't it be a generic that defaults to any? So you can do `request<BodyType>`?

Reply:
This makes sense. I can make that change.